src/Pirouette/Term/TypeChecker.hs

```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+module Pirouette.Term.TypeChecker where
+
+import Data.Bifunctor
+import qualified Data.Map as Map
+import Control.Monad.Except
+import Control.Monad.Reader
+import Pirouette.Term.Syntax.Base hiding (Var, TyVar)
+import Pirouette.Term.Syntax.SystemF
+import Pirouette.Term.Syntax.Subst (shift)
+
+-- | Checks an entire set of declarations.
+typeCheckDecls
+  :: LanguageBuiltinTypes lang
+  => Decls lang -> Either (TypeError lang) ()
+typeCheckDecls decls = do
+  _ <- flip Map.traverseWithKey decls $ \name decl -> case decl of
+    DFunDef FunDef { funBody, funTy } ->
+      flip runReaderT ((decls, []), [DeclPath name]) $
+        void $ typeCheckTerm funTy funBody
+    _ -> pure ()
+  pure ()
+
+class (LanguageBuiltins lang)
+      => LanguageBuiltinTypes lang where
+  typeOfConstant :: Constants lang    -> Type lang
+  typeOfBuiltin  :: BuiltinTerms lang -> Type lang
+  typeOfBottom   :: Type lang
```

Comment:
Should we add this class to `Pirouette.Term.Syntax.Base`?

Reply:
I guess both keeping it here (because it only applies to type checking) and moving it to base (because it's a quite common thing) make sense. I don't have a strong opinion on this.

Reply:
Moved there, this was nice because now some modules don't have to depend on the `TypeChecker` one.